File: 0900-rle-iterator/0900-rle-iterator.cpp
class RLEIterator {
    vector<int> data;
    int cur_idx = 0;
public:
    RLEIterator(vector<int>& encoding) {
        data = encoding;
    }
    
    int next(int n) {
        if(cur_idx >= data.size())  return -1;

        if(data[cur_idx] >= n){
            data[cur_idx] -= n;

            int val = data[cur_idx+1];
            while(cur_idx < data.size() && data[cur_idx] == 0)  cur_idx += 2;
            return val;
        }
        else{
            n -= data[cur_idx];
            cur_idx += 2;
            return next(n);
        }
    }
};

/**
 * Your RLEIterator object will be instantiated and called as such:
 * RLEIterator* obj = new RLEIterator(encoding);
 * int param_1 = obj->next(n);
 */

File: 0900-rle-iterator/0900-rle-iterator.py
class RLEIterator:

    def __init__(self, encoding: List[int]):
        self.encoded = encoding[:]
        

    def next(self, n: int) -> int:
        
        count = 0
        
        while count < n:
            if self.encoded:
                no_of_times = self.encoded.pop(0)
                element = self.encoded.pop(0)
                
                count += no_of_times
            else:
                return -1
        
        if count > n:
            self.encoded.insert(0, element)
            self.encoded.insert(0, count-n)
        
        return element
# Your RLEIterator object will be instantiated and called as such:
# obj = RLEIterator(encoding)
# param_1 = obj.next(n)